src/LinuxTracing/LinuxTracingIntegrationTest.cpp

```diff
@@ -0,0 +1,301 @@
+  }
+
+  void OnTracepointEvent(orbit_grpc_protos::TracepointEvent tracepoint_event) override {
+    orbit_grpc_protos::CaptureEvent event;
+    *event.mutable_tracepoint_event() = std::move(tracepoint_event);
+    events_.emplace_back(std::move(event));
+  }
+
+  void OnModulesUpdate(orbit_grpc_protos::ModuleUpdateEvent module_update_event) override {
+    orbit_grpc_protos::CaptureEvent event;
+    *event.mutable_module_update_event() = std::move(module_update_event);
+    events_.emplace_back(std::move(event));
+  }
+
+  [[nodiscard]] const std::vector<orbit_grpc_protos::CaptureEvent>& GetEvents() const {
+    return events_;
+  }
+
+  [[nodiscard]] std::vector<orbit_grpc_protos::CaptureEvent>&& GetEvents() {
+    return std::move(events_);
+  }
+
+ private:
+  std::vector<orbit_grpc_protos::CaptureEvent> events_;
+};
+
+// GTest's test fixtures seem to mess with the pipe logic in ChildProcess.
+// Let's make the fixture ourselves, which is also advised in http://go/totw/122.
+class LinuxTracingIntegrationTestFixture {
+ public:
+  LinuxTracingIntegrationTestFixture() : puppet_{&LinuxTracingIntegrationTestPuppetMain} {}
+
+  [[nodiscard]] pid_t GetPuppetPid() const { return puppet_.GetChildPid(); }
+
+  void WriteLineToPuppet(std::string_view str) { puppet_.WriteLine(str); }
+
+  [[nodiscard]] std::string ReadLineFromPuppet() { return puppet_.ReadLine(); }
+
+  void StartTracing() {
+    CHECK(!tracer_.has_value());
+    CHECK(!listener_.has_value());
+    orbit_grpc_protos::CaptureOptions capture_options = BuildCaptureOptions();
+    tracer_.emplace(capture_options);
+    listener_.emplace();
+    tracer_->SetListener(&*listener_);
+    tracer_->Start();
+  }
+
+  [[nodiscard]] std::vector<orbit_grpc_protos::CaptureEvent> StopTracingAndGetEvents() {
+    CHECK(tracer_.has_value());
+    CHECK(listener_.has_value());
+    tracer_->Stop();
+    tracer_.reset();
+    std::vector<orbit_grpc_protos::CaptureEvent> events = std::move(listener_->GetEvents());
+    listener_.reset();
+    return events;
+  }
+
+ private:
+  [[nodiscard]] orbit_grpc_protos::CaptureOptions BuildCaptureOptions() {
+    orbit_grpc_protos::CaptureOptions capture_options;
+    capture_options.set_trace_context_switches(true);
+    capture_options.set_pid(puppet_.GetChildPid());
+    capture_options.set_sampling_rate(1000.0);
+    capture_options.set_unwinding_method(orbit_grpc_protos::CaptureOptions::kDwarf);
+    capture_options.set_trace_thread_state(true);
+    capture_options.set_trace_gpu_driver(true);
+    return capture_options;
+  }
+
+ private:
+  ChildProcess puppet_;
+  std::optional<Tracer> tracer_ = std::nullopt;
+  std::optional<BufferTracerListener> listener_ = std::nullopt;
+};
+
+void SleepMs(uint64_t ms) { std::this_thread::sleep_for(std::chrono::milliseconds{ms}); }
+
+using PuppetConstants = LinuxTracingIntegrationTestPuppetConstants;
+
+TEST(LinuxTracingIntegrationTest, ModuleUpdateOnDlopen) {
+  if (!CheckIsPerfEventParanoidAtMost(0)) GTEST_SKIP();
```

Comment:
<< "This test should be run under root or with perf event paranoid value at most 0"

Reply:
also {}

Reply:
That doesn't print the message. `CheckIsPerfEventParanoidAtMost` prints a message, though.